core/src/test/java/com/linecorp/armeria/client/InitiateConnectionShutdownTest.java

```diff
@@ -0,0 +1,229 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.awaitility.Awaitility.await;
+
+import java.net.InetSocketAddress;
+import java.time.Duration;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvSource;
+
+import com.google.common.collect.Lists;
+
+import com.linecorp.armeria.common.AggregatedHttpResponse;
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.common.logging.RequestLogAccess;
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.testing.junit5.server.ServerExtension;
+
+import io.netty.handler.codec.http.HttpHeaderValues;
+import io.netty.util.AttributeMap;
+
+public class InitiateConnectionShutdownTest {
+    @RegisterExtension
+    static final ServerExtension server = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) throws Exception {
+            sb.service("/ok", (ctx, req) -> HttpResponse.of(HttpStatus.OK));
+            sb.service("/delayed_ok", (ctx, req) -> {
+                return HttpResponse.delayed(HttpResponse.of(HttpStatus.OK),
+                                            Duration.ofMillis(500));
+            });
+        }
+    };
+
+    final List<RequestLogAccess> requestLogAccesses = Lists.newCopyOnWriteArrayList();
+    final AtomicBoolean completed = new AtomicBoolean();
+    final AtomicInteger connectionOpen = new AtomicInteger();
+    final AtomicInteger connectionClosed = new AtomicInteger();
+    final ConnectionPoolListener poolListener = new ConnectionPoolListener() {
+        @Override
+        public void connectionOpen(SessionProtocol protocol, InetSocketAddress remoteAddr,
+                                   InetSocketAddress localAddr, AttributeMap attrs) throws Exception {
+            connectionOpen.incrementAndGet();
+        }
+
+        @Override
+        public void connectionClosed(SessionProtocol protocol, InetSocketAddress remoteAddr,
+                                     InetSocketAddress localAddr, AttributeMap attrs) throws Exception {
+            connectionClosed.incrementAndGet();
+        }
```

Comment:
nit: How about using `CountingConnectionPoolListener` which was added for the testing purpose?